### PR TITLE
Gangams/logs multi tenancy

### PR DESCRIPTION
**Pull request by ganga1980**
This PR implements Logs Multi-tenancy feature which enables the  ingesting of one or more k8s namespace container logs to corresponding log analytics workspace destination. With this, for example logs of app-team-1 can ingested to app-team-1 log analytics workspace, logs of app-team-2 can ingested to app-team-2 log analytics workspace and logs of infra team can be ingested to infra-team log analytics workspace etc. 

Pre-requisite of this feature to enable high log scale mode. As part of this feature ContainerLogV2 extension DCR, which created which can be created N of DCR, each DCR with unique azure log analytics destination and data collection settings. 

**Tested both Linux and Windows scenarios.**

### CODE REVIEW

**Comment by ganga1980** [comment]
verified with 31 DCR ( default DCR and 30 CLv2 extension DCRs) and each  CLv2 DCR with one app-namespace, test workload generating 2k log/sec per each namespace (i.e. 60k logs/sec/node for app namespaces) and total logs generated by each pod in each namespace is100K log lines. Default DCR for the system namespaces.  Here are the various charts

image
image
image
Logs/per minute by computer 

image

File: Documentation/DCR/ContainerLogV2Extension.json [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+                    "name": "la-workspace"
+                }
+            ]
+        },
+        "dataFlows": [
+            {
+                "streams": [
+                    "Microsoft-ContainerLogV2-HighScale"
+                ],
+                "transformKql": "source | where LogLevel in ('CRITICAL', 'ERROR')",
```
Comment (wanlonghenry) [inline Documentation/DCR/ContainerLogV2Extension.json:33]: can the loglevel be made configurable for kql transform

File: build/common/installer/scripts/tomlparser.rb [hunk 2]
```diff
@@ -322,6 +323,16 @@ def populateSettingValuesFromConfigMap(parsedConfig)
     rescue => errorStr
       ConfigParseErrorLogger.logError("config::error: Exception while reading config map settings for annotation based log filtering - #{errorStr}, please check config map for errors")
     end
+
+    #Get Multi-tenancy log collection settings
+    begin
+      if !parsedConfig[:log_collection_settings][:multi_tenancy].nil? && !parsedConfig[:log_collection_settings][:multi_tenancy][:enabled].nil?
```
Comment (wanlonghenry) [inline build/common/installer/scripts/tomlparser.rb]: I don't see this added in configmap, will this be added in future stage?  Also should the name follow multi_tenancy_log_collection

File: scripts/onboarding/aks/multi-tenancy/existingClusterOnboarding.json [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+                            "location": "[parameters('workspaceRegion')]",
+                            "tags": "[parameters('resourceTagValues')]",
+                            "kind": "Linux",
+                            "properties": {
+                                "dataSources": {
+                                    "extensions": [
+                                        {
+                                            "name": "ContainerLogV2Extension",
+                                            "streams": [
+                                                "Microsoft-ContainerLogV2-HighScale"
```
Comment (wanlonghenry) [inline scripts/onboarding/aks/multi-tenancy/existingClusterOnboarding.json:142]: will this be support stream configuration in future

File: scripts/onboarding/aks/multi-tenancy/existingClusterOnboarding.json [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+                                            ],
+                                            "extensionSettings": {
+                                                "dataCollectionSettings": {
+                                                    "namespaces": "[parameters('k8sNamespaces')]"
```
Comment (wanlonghenry) [inline scripts/onboarding/aks/multi-tenancy/existingClusterOnboarding.json:146]: how about
"interval": "[parameters('dataCollectionInterval')]",
                                            "namespaceFilteringMode": "[parameters('namespaceFilteringModeForDataCollection')]",
                                            "namespaces": "[parameters('namespacesForDataCollection')]",
                                            "enableContainerLogV2": "[parameters('enableContainerLogV2')]"

also here we should set enableContainerLogV2 as true as default right

File: scripts/onboarding/aks/multi-tenancy/existingClusterOnboarding.json [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+            "properties": {
+                "networkAcls": {
+                    "publicNetworkAccess": "Enabled"
+                }
+            }
+        },
+        {
+            "type": "Microsoft.Resources/deployments",
+            "name": "[Concat('aks-monitoring-msi-dcr', '-',  uniqueString(variables('dcrName')))]",
+            "apiVersion": "2017-05-10",
```
Comment (wanlonghenry) [inline scripts/onboarding/aks/multi-tenancy/existingClusterOnboarding.json:115]: do we want to use new api version?

File: source/plugins/go/src/extension/extension.go [hunk 6]
```diff
@@ -320,6 +320,91 @@ func (e *Extension) GetNamespaceFilteringModeForDataCollection() string {
 	return namespaceFilteringMode
 }
 
+func (e *Extension) GetContainerLogV2ExtensionConfig(isWindows bool) (map[string][]string, map[string]string, error) {
+	namespaceStreamIdsMap := make(map[string][]string)
+	StreamIdNamedPipeMap := make(map[string]string)
```
Comment (wanlonghenry) [inline source/plugins/go/src/extension/extension.go]: camelCase

File: source/plugins/go/src/extension/extension.go [hunk 7]
```diff
@@ -320,6 +320,91 @@ func (e *Extension) GetNamespaceFilteringModeForDataCollection() string {
...
+
+	err = json.Unmarshal([]byte(responseObject.TaggedData), &extensionData)
+	extensionConfigs := extensionData.ExtensionConfigs
+	outputStreamDefinitions := make(map[string]StreamDefinition)
+	if isWindows {
+		outputStreamDefinitions = extensionData.OutputStreamDefinitions
+	}
+
+	for _, extensionConfig := range extensionConfigs {
+		outputStreamId := ""
```
Comment (wanlonghenry) [inline source/plugins/go/src/extension/extension.go:359]: outputStreamID?

File: source/plugins/go/src/extension/extension.go [hunk 8]
```diff
@@ -320,6 +320,91 @@ func (e *Extension) GetNamespaceFilteringModeForDataCollection() string {
...
+				} else {
+					logger.Printf("GetContainerLogV2ExtensionConfig:: outputStreamId: %s", outputStreamId)
+				}
+			}
+		}
+
+		extensionSettings := extensionConfig.ExtensionSettings
+		dataCollectionSettingsItr, ok := extensionSettings["dataCollectionSettings"]
+
+		dataCollectionSettings := make(map[string]interface{})
```
Comment (wanlonghenry) [inline source/plugins/go/src/extension/extension.go]: extension.go has a getDataCollectionSettingsInterface method which do the similar thing

File: source/plugins/go/src/oms.go [hunk 9]
```diff
@@ -475,6 +489,40 @@ func updateContainerImageNameMaps() {
 	}
 }
 
+func updateContainerLogV2ExtensionMaps(isWindows bool) {
+	for ; true; <-ContainerLogV2ExtensionConfigRefreshTicker.C {
```
Comment (wanlonghenry) [inline source/plugins/go/src/oms.go:493]: for range ContainerLogV2ExtensionConfigRefreshTicker.C

File: source/plugins/go/src/oms.go [hunk 10]
```diff
@@ -475,6 +489,40 @@ func updateContainerImageNameMaps() {
...
+				Log("updateContainerLogV2ExtensionMaps::error: %s attempt: %d", string(err.Error()), attempt)
+				time.Sleep(time.Duration(attempt+1) * time.Second)
+			} else {
+				Log("updateContainerLogV2ExtensionMaps:Info:Locking to update NamespaceStreamIdsMap and StreamIdNamedPipeMap")
+				ContainerLogV2ExtensionMapUpdateMutex.Lock()
+				for key := range NamespaceStreamIdsMap {
+					delete(NamespaceStreamIdsMap, key)
+				}
+				for key, value := range _namespaceStreamIdsMap {
+					NamespaceStreamIdsMap[key] = value
```
Comment (wanlonghenry) [inline source/plugins/go/src/oms.go:508]: can this be replaced with direct assign? NamespaceStreamIdsMap = _namespaceStreamIdsMap

File: source/plugins/go/src/oms.go [hunk 11]
```diff
@@ -475,6 +489,40 @@ func updateContainerImageNameMaps() {
...
+				}
+				if isWindows {
+					for key := range StreamIdNamedPipeMap {
+						delete(StreamIdNamedPipeMap, key)
+					}
+					for key, np := range _streamIdNamedPipeMap {
+						StreamIdNamedPipeMap[key] = np
```
Comment (wanlonghenry) [inline source/plugins/go/src/oms.go:515]: same might be replaced with direct assign

File: source/plugins/go/src/oms.go [hunk 12]
```diff
@@ -1880,6 +1896,128 @@ func containsKey(currentMap map[string]bool, key string) bool {
...
+						deadline := 10 * time.Second
+						if IsWindows {
+							// in windows, there will be dedicated namedpipe for each DCR and hence use namedpipe specific to DCR
+							namedPipe, ok := streamIdNamedPipeMap[streamTag]
+							Log("Info::ama:: namedpipe: %s and streamTag: %s: namespace: %s \n", namedPipe, streamTag, namespace)
+							if ok {
+								namedPipeConn, ok := NamedPipeConnectionCache[namedPipe]
+								if !ok || namedPipeConn == nil {
+									CreateWindowsNamedPipeClient(namedPipe, &namedPipeConn)
+									NamedPipeConnectionCache[namedPipe] = namedPipeConn
```
Comment (wanlonghenry) [inline source/plugins/go/src/oms.go:1932]: should this cache be added after checking write success?

File: build/common/installer/scripts/tomlparser.rb (hunk 2 above)
Comment (ganga1980) [reply build/common/installer/scripts/tomlparser.rb]: log collection is already top level setting hence used the child config property as multi_tenancy.

File: scripts/onboarding/aks/multi-tenancy/existingClusterOnboarding.json (hunk 3 above)
Comment (ganga1980) [reply scripts/onboarding/aks/multi-tenancy/existingClusterOnboarding.json:142]: > will this be support stream configuration in future

what do you mean stream configuration?

File: scripts/onboarding/aks/multi-tenancy/existingClusterOnboarding.json (hunk 4 above)
Comment (ganga1980) [reply scripts/onboarding/aks/multi-tenancy/existingClusterOnboarding.json:146]: all these settings are not applicable . like kubernetes metadata and annotation feature, containerlogv2 is dependency for this feature.

File: source/plugins/go/src/oms.go (hunk 9 above)
Comment (ganga1980) [reply source/plugins/go/src/oms.go:493]: no, existing code correct.

File: source/plugins/go/src/extension/extension.go (hunk 6 above)
Comment (ganga1980) [reply source/plugins/go/src/extension/extension.go]: > camelCase

good catch. Fixed.

File: source/plugins/go/src/extension/extension.go (hunk 7 above)
Comment (ganga1980) [reply source/plugins/go/src/extension/extension.go:359]: > outputStreamID?

there is already variable with that hence used this to differentiate.

File: source/plugins/go/src/oms.go (hunk 10 above)
Comment (ganga1980) [reply source/plugins/go/src/oms.go:508]: > can this be replaced with direct assign? NamespaceStreamIdsMap = _namespaceStreamIdsMap

No, doing  this NamespaceStreamIdsMap = _namespaceStreamIdsMap will use the same reference but not the contents. we would need only contents need to be copied.

File: source/plugins/go/src/oms.go (hunk 12 above)
Comment (ganga1980) [reply source/plugins/go/src/oms.go:1932]: good catch. fixed.

File: source/plugins/go/src/oms.go (hunk 11 above)
Comment (ganga1980) [reply source/plugins/go/src/oms.go:515]: No, direct assign copy the ref not value. we would need contents to be copied.

File: scripts/onboarding/aks/multi-tenancy/existingClusterOnboarding.json (hunk 3 above)
Comment (wanlonghenry) [reply scripts/onboarding/aks/multi-tenancy/existingClusterOnboarding.json:142]: like "streams": {
      "value": [
        "Microsoft-ContainerLog",
        "Microsoft-ContainerLogV2",
        "Microsoft-KubeEvents",
        "Microsoft-KubePodInventory",
        "Microsoft-KubeNodeInventory",
        "Microsoft-KubePVInventory",
        "Microsoft-KubeServices",
        "Microsoft-KubeMonAgentEvents",
        "Microsoft-InsightsMetrics",
        "Microsoft-ContainerInventory",
        "Microsoft-ContainerNodeInventory",
        "Microsoft-Perf"
      ]
    }

Comment (ganga1980) [reply scripts/onboarding/aks/multi-tenancy/existingClusterOnboarding.json:142]: No, only supported stream with this DCR is Microsoft-ContainerLogV2-HighScale

File: source/plugins/go/src/extension/extension.go (hunk 8 above)
Comment (ganga1980) [reply source/plugins/go/src/extension/extension.go]: > extension.go has a getDataCollectionSettingsInterface method which do the similar thing

refactored.

File: scripts/onboarding/aks/multi-tenancy/existingClusterOnboarding.json (hunk 5 above)
Comment (ganga1980) [reply scripts/onboarding/aks/multi-tenancy/existingClusterOnboarding.json:115]: > do we want to use new api version?

update the api version for DCR, DCE and DCR-A to latest version.

**Pull request closed without merging**